add air attenuation after iso

ahms5 opened the pull request:
- [x] implement air attenuataion after ISO
- [x] separate vapour pressure calcualtion into private function (there is just a sligt difference between this and the magnus formula in #770) 
![grafik](https://github.com/user-attachments/assets/47e7ec54-5640-4aed-9d48-4628745301e3)

ahms5 commented on pyfar/constants/air_attenuation.py:
```diff
@@ -0,0 +1,243 @@
+    return air_attenuation, accuracy
+
+
+def _p_sat_water(temperature):
```
this is was replaced by saturation_vapor_pressure and is not used anymore, its just temporary here and is comping from #503

f-brinkmann commented on pyfar/constants/air_attenuation.py:
```diff
@@ -0,0 +1,243 @@
+import numpy as np
+
+
+def air_attenuation_iso(
```
Is there another air attenuation that we want to implement? Not sure if we need the ISO in the name. In other places, we also don't mention the source in the function name.

f-brinkmann commented on pyfar/constants/air_attenuation.py:
```diff
@@ -0,0 +1,243 @@
+        raise ValueError("Atmospheric pressure must less than 200 kPa.")
+
+    # convert arrays
+    all_are_singles = isinstance(temperature, (float, int))
+    all_are_singles = all_are_singles and isinstance(frequency, (float, int))
+    all_are_singles = all_are_singles and isinstance(
+        relative_humidity, (float, int))
+    if not all_are_singles:
+        shape = np.broadcast_shapes(
+            np.array(temperature).shape,
+            np.array(frequency).shape,
+            np.array(relative_humidity).shape,
+            np.array(atmospheric_pressure).shape)
+        temperature = np.broadcast_to(
+            np.array(temperature, dtype=float), shape)
+        frequency = np.broadcast_to(
+            np.array(frequency, dtype=float), shape)
+        relative_humidity = np.broadcast_to(
+            np.array(relative_humidity, dtype=float), shape)
+        atmospheric_pressure = np.broadcast_to(
+            np.array(atmospheric_pressure, dtype=float), shape)
```
Is the explicit broadcasting needed? I would suggest to convert input to numpy arrays in any case and maybe only
```
try:
    _ = np.broadcast_shapes(...)
except: ValueError as e:
    raise ValueError('some message') from e
```

ahms5 replied:
yes for the accuracy later on

f-brinkmann commented on pyfar/constants/air_attenuation.py:
```diff
@@ -0,0 +1,243 @@
+
+    Returns
+    -------
+    air_attenuation : float, array_like
+        Pure tone air attenuation coefficient in decibels per meter for
+        atmospheric absorption.
+    accuracy : float, array_like, optional
```
I would suggest to return both as FrequencyData objects

f-brinkmann replied:
I think the accuracy is not tested.

f-brinkmann commented on pyproject.toml:
```diff
@@ -32,7 +32,7 @@ classifiers = [
     "numpy>=1.23.0",
     "scipy>=1.5.0",
-    "matplotlib",
+    "matplotlib<3.10",
```
We could discuss, which solution we want to use...

f-brinkmann reviewed (changes requested):
Thanks for implementing. Looks mostly fine to me. I checked against our Matlab implementation. Results generally agree but are slightly different. I think this is related to the improved Magnus Formula, which we didn't implement. One general question:

Do we want to add a flag to return the air attenuation in dB or as energetic property? The latter is used in common approximations of the reverberation time, and might be useful as well. I would be in favor of this.

ahms5 commented on pyproject.toml:
```diff
@@ -32,7 +32,7 @@ classifiers = [
     "numpy>=1.23.0",
     "scipy>=1.5.0",
-    "matplotlib",
+    "matplotlib<3.10",
```
```suggestion
    "matplotlib",
```

ahms5 commented:
In the standard, there is a table of results that I use for comparison during the tests. The Magnus formula in the ITA-Toolbox and AKtools was not well documented, as it referenced the manual of a humidity measuring device. Therefore, I searched for a more reliable reference.

f-brinkmann reviewed (changes requested):
Thanks for the effort the computation and testing looks good to me - I checked it against our Matlab implementation from AKtools. But there might be an error when computing the accuracy. For `alpha, m, aaccuracy = pf.constants.air_attenuation(20, 1e3, .5, 101325)` I get `-1` for the accuracy, where it shoud be `10` I think.

ahms5 commented on pyfar/constants/air_attenuation.py:
```diff
@@ -0,0 +1,184 @@
+    -------
+    alpha :  :py:class:`~pyfar.classes.FrequencyData`
+        Pure tone air attenuation coefficient in decibels per meter for
+        atmospheric absorption. The cshape of the output is the broadcast
```
its a FrequencyData object, do we need to explain that anywhere. I could add it into the function describtion, what do you think?

sikersten commented on pyfar/constants/air_attenuation.py:
```diff
@@ -0,0 +1,184 @@
+        Frequency in Hz. Must be greater than 50 Hz.
+        Just one dimensional array is allowed.
+    relative_humidity : float, array_like
+        Relative humidity in the range of 0 to 1.
```
```suggestion
        Relative humidity in the range from 0 to 1.
```

sikersten commented on pyfar/constants/air_attenuation.py:
```diff
@@ -0,0 +1,184 @@
+    accuracy = np.zeros(shape) - 1
+    atmospheric_pressure = np.broadcast_to(atmospheric_pressure, shape)
+    p_vapor = np.broadcast_to(p_vapor, shape)/100
+    freq2pressure = frequencies/atmospheric_pressure
```
```suggestion
    freq_pressure_ratio = frequencies/atmospheric_pressure
```

sikersten commented on pyfar/constants/air_attenuation.py:
```diff
@@ -0,0 +1,184 @@
+    m = alpha / (10*np.log10(np.exp(1)))
+
+    # calculate accuracy
+    accuracy = np.zeros(shape) - 1
```
I wonder if there a good alternatives to -1, e.g., nan, None, 0?

sikersten commented on pyfar/constants/air_attenuation.py:
```diff
@@ -0,0 +1,184 @@
+            - molar concentration of water vapour: 0,005 % to 0,05 %,
+              and greater than 5%
+            - air temperature: 253,15 K to 323,15 (-20 °C to +50°C)
+            - air temperature: greater than 200 K (- 73 °C)
```
I think the standard states -20 to 50 for the 20% accuracy

sikersten commented on pyfar/constants/air_attenuation.py:
```diff
@@ -0,0 +1,184 @@
+    accuracy[accuracy_10] = 10
+
+    vapor_2_mask = (vapor_1_mask is False) & (0.005 <= p_vapor)
+    temp_2_mask = (-70 <= temperature) & (temp_1_mask is False)
```
Sure about this temperature range for 20% accuracy?

sikersten replied:
I think this needs to be -73 instead of -70

sikersten commented on tests/test_constants_air_attenuation.py:
```diff
@@ -0,0 +1,116 @@
+    (10, 100, .1, 5.85e-1*1e-3, 10),
+    (10, 100, .005, 0.002094, 10),
+])
+def test_air_attenuation_iso(
```
```suggestion
def test_air_attenuation(
```

sikersten commented on tests/test_constants_air_attenuation.py:
```diff
@@ -0,0 +1,116 @@
+        [101325, 101325],
+        101325,
+])
+def test_air_attenuation_iso_array(
```
```suggestion
def test_air_attenuation_array(
```

sikersten commented on tests/test_constants_air_attenuation.py:
```diff
@@ -0,0 +1,116 @@
+    npt.assert_allclose(accuracy.frequencies, frequency, atol=1e-3)
+
+
+def test_air_attenuation_iso_inputs():
```
```suggestion
def test_air_attenuation_inputs():
```

sikersten commented on tests/test_constants_air_attenuation.py:
```diff
@@ -0,0 +1,116 @@
+            temperature, frequency, relative_humidity, 'test')
+
+
+def test_air_attenuation_iso_broadcastable():
```
```suggestion
def test_air_attenuation_broadcastable():
```

sikersten commented on tests/test_constants_air_attenuation.py:
```diff
@@ -0,0 +1,116 @@
+            [10, 10], frequency, [.1, .1, .1])
+
+
+def test_air_attenuation_iso_limits():
```
```suggestion
def test_air_attenuation_limits():
```

sikersten commented on tests/test_constants_air_attenuation.py:
```diff
@@ -0,0 +1,116 @@
+
+
+@pytest.mark.parametrize((
+        "temperature", "frequency", "relative_humidity", "expected",
```
How did you came up with the expected values?

sikersten commented on pyfar/constants/medium_attenuation.py:
```diff
@@ -0,0 +1,254 @@
+    common_mask = atm_mask & freq2pressure_mask
+
+    # molar concentration of water vapour: 0% to 100%
+    vapor_com = (0 <= h_water_vapor) & (h_water_vapor <= 100)
```
What do you think of replacing this with an error message, if the value is outside 0-100?

sikersten replied:
This would also allow removing ``& vapor_com`` everywhere below

sikersten commented on pyfar/constants/medium_attenuation.py:
```diff
@@ -0,0 +1,254 @@
+
+    # molar concentration of water vapour: 0% to 100%
+    vapor_com = (0 <= h_water_vapor) & (h_water_vapor <= 100)
+    # molar concentration of water vapour: 0.005% to 0.05 % and greater than 5%
+    vapor_1_mask = (0.05 <= h_water_vapor) & (h_water_vapor <= 5) & vapor_com
+    # molar concentration of water vapour: 0.05% to 5 %
+    vapor_2_mask = ((5 < h_water_vapor) | (h_water_vapor >= 0.005)) & vapor_com
+    vapor_2_mask = vapor_2_mask | vapor_1_mask
+    # molar concentration of water vapour: less than 0.005%
+    vapor_3_mask = ((0.005 > h_water_vapor)  & vapor_com) | vapor_2_mask
```
```suggestion
    # molar concentration of water vapour: 0.05% to 5 %
    vapor_1_mask = (0.05 <= h_water_vapor) & (h_water_vapor <= 5) & vapor_com
    # molar concentration of water vapour: 0.005% to 0.05 %, or greater than 5%
    vapor_2_mask = (((0.005 <= h_water_vapor) & (h_water_vapor < 0.05)) | (5 < h_water_vapor)) & vapor_com
    # molar concentration of water vapour: less than 0.005%
    vapor_3_mask = (h_water_vapor < 0.005)  & vapor_com
```

sikersten commented on tests/test_constants_air_attenuation.py:
```diff
@@ -0,0 +1,208 @@
+        temperature, frequency, relative_humidity, expected,
+        expected_accuracy):
+    """
+    Check alpha after value table in the standard. The accuracy is calculated
```
```suggestion
    Check alpha after value table in ISO 9613-1. The accuracy is calculated
```

ahms5 commented on pyfar/constants/speed_of_sound.py:
```diff
@@ -40,6 +41,94 @@ def speed_of_sound_simple(temperature):
     return 343.2*np.sqrt((temperature+273.15)/(t_0+273.15))
+
+
+def speed_of_sound_ideal_gas(
```
this file was already reviewed in #739

ahms5 commented:
@CristobalAndrade pointed out that ISO Appendix B provides a defined method for calculating saturation vapor pressure. Therefore, I removed the old saturation_vapor_pressure function (still required for ideal gas in #770) and incorporated the calculation directly into the air attenuation function, following the standard's guidelines.

ahms5 commented on tests/test_constants_air_attenuation.py:
```diff
@@ -0,0 +1,247 @@
+    with pytest.raises(ValueError, match=match):
+        pf.constants.air_attenuation(
+            temperature, frequency, 1.1, atmospheric_pressure)
+    match = 'Atmospheric pressure must less than 200 kPa.'
```
```suggestion
    match = 'Atmospheric pressure must be less than 200 kPa.'
```

sikersten commented on pyfar/constants/constants.py:
```diff
@@ -0,0 +1,302 @@
+    relative_humidity : float, array_like
+        Relative humidity in the range from 0 to 1.
+    atmospheric_pressure : float, array_like, optional
+        Atmospheric pressure in pascal, by default
```
```suggestion
        Atmospheric pressure in Pascal, by default
```

sikersten commented on pyfar/constants/constants.py:
```diff
@@ -0,0 +1,302 @@
+        Temperature in degree Celsius.
+        Must be above -273.15°C.
+    atmospheric_pressure : float, array_like
+        Atmospheric pressure in pascal.
```
```suggestion
        Atmospheric pressure in Pascal.
```

sikersten commented on pyfar/constants/constants.py:
```diff
@@ -0,0 +1,302 @@
+    Returns
+    -------
+    p_sat : float, array_like
+        Saturation vapour pressure in pascal.
```
```suggestion
        Saturation vapour pressure in Pascal.
```

sikersten commented on pyfar/constants/constants.py:
```diff
@@ -0,0 +1,302 @@
+            - molar concentration of water vapour: 0.05% to 5 %.
+            - air temperature: 253.15 K to 323.15 (-20 °C to +50°C)
+            - atmospheric pressure: less than 200 000 Pa (2 atm)
+            - frequency-to-pressure ratio: 4 x 10-4 Hz/Pa to 10 Hz/Pa.
```
Is there an easy solution to put the "-4" as exponent?

sikersten commented on pyfar/constants/constants.py:
```diff
@@ -0,0 +1,302 @@
+    alpha = pf.FrequencyData(
+        air_attenuation, frequencies=frequencies)
+
+    m = alpha / (10*np.log10(np.exp(1)))
```
The documentation states a multiplication instead of division. Is there a reference for this equation?

sikersten commented on pyfar/constants/constants.py:
```diff
@@ -0,0 +1,302 @@
+    m : :py:class:`~pyfar.classes.FrequencyData`
+        Pure tone air attenuation coefficient per meter for
+        atmospheric absorption. The parameter ``m`` is calculated as
+        :math:`m = 10 \cdot \log(\exp(1)) \cdot \alpha`.
```
Differs from the actual implementation, see below.

sikersten commented on pyfar/constants/constants.py:
```diff
@@ -0,0 +1,302 @@
+        Must be above 0Pa.
+    frequencies : float, array_like
+        Frequency in Hz.
+        Must be lager than 0 Hz.
```
```suggestion
        Must be larger than 0 Hz.
```

sikersten commented on pyfar/constants/constants.py:
```diff
@@ -0,0 +1,302 @@
+def _calculate_accuracy(
+        concentration_water_vapour, temperature, atmospheric_pressure,
+        frequencies, shape):
+    """Calculate the accuracy of the air attenuation calculation.
```
It would make sense to mention the standard somewhere in the docstring

mberz commented on pyfar/constants/constants.py:
```diff
@@ -0,0 +1,313 @@
+    return alpha, m, accuracy
+
+
+def _calculate_accuracy(
```
Accuracy of what? Please be more concise.

mberz reviewed (changes requested):
Only one comment...

.... but why does this PR have 55 commits? 🤣
